recipes/full_finetune_single_device.py

```diff
@@ -221,18 +229,55 @@ def _setup_model(
         return model
 
     def _setup_optimizer(
-        self, cfg_optimizer: DictConfig, opt_state_dict: Optional[Dict[str, Any]] = None
-    ) -> Optimizer:
+        self,
+        cfg_optimizer: DictConfig,
+        optimizer_in_bwd: bool = False,
+        opt_state_dict: Optional[Dict[str, Any]] = None,
+    ) -> Union[Optimizer, None]:
         """
         Set up the optimizer. This method also handles loading the optimizer state_dict, if specified.
         """
-        optimizer = config.instantiate(cfg_optimizer, self._model.parameters())
-
-        if opt_state_dict:
-            optimizer.load_state_dict(opt_state_dict)
+        if optimizer_in_bwd:
+            self._optimizer_in_bwd = True
+            # Maintain a dict of optims for every parameter.
+            # TODO (rohan-varma): check foreach arg
+            optim_dict = {
+                p: config.instantiate(cfg_optimizer, [p])
+                for p in self._model.parameters()
+            }
+            self._optim_ckpt_wrapper = utils.OptimizerInBackwardWrapper(
+                {n: optim_dict[p] for n, p in self._model.named_parameters()}
+            )
 
-        log.info("Optimizer is initialized.")
-        return optimizer
+            def optim_step(param) -> None:
+                optim_dict[param].step()
+                optim_dict[param].zero_grad()
+
+            for p in self._model.parameters():
+                p.register_post_accumulate_grad_hook(optim_step)
+
+            # Load optimizer states. If optimizer states are being restored in an optimizer in backward
+            # run, these need to have been saved with the same setting. Cannot restore from runs that did not
+            # use optimizer in backward.
+            if opt_state_dict is not None:
+                try:
+                    self._optim_ckpt_wrapper.load_state_dict(opt_state_dict)
+                except BaseException as e:
+                    raise RuntimeError(
+                        "Failed loading in-backward optimizer checkpoints."
+                        "Please make sure run being restored from was using in-backward optimizer."
+                        f"Original error {str(e)}"
```

Comment:
the `from e` should take care of surfacing the original error

Reply:
I actually tried this, and it didn't - i.e. I didn't see the exception from `e` in this error.

Reply:
Let me continue checking this though